Mathlib/Tactic/Module.lean

```diff
@@ -0,0 +1,641 @@
+and a natural number), construct another such term `l`, which will have the property that in the
+`$R`-module `$M`, the difference of the "linear combinations" represented by `l₁` and `l₂` is the
+linear combination represented by `l`.
+
+The construction assumes, to be valid, that the lists `l₁` and `l₂` are in strictly increasing order
+by `ℕ`-component, and that if pairs `(a₁, x₁)` and `(a₂, x₂)` appear in `l₁`, `l₂` respectively with
+the same `ℕ`-component `k`, then the expressions `x₁` and `x₂` are equal.
+
+The construction is as follows: merge the first list and the negation of the second list, except
+that if pairs `(a₁, x₁)` and `(a₂, x₂)` appear in `l₁`, `l₂` respectively with the same
+`ℕ`-component `k`, then contribute a term `(a₁ - a₂, x₁)` to the output list with `ℕ`-component `k`.
+-/
+def sub (iR : Q(Ring $R)) : qNF R M → qNF R M → qNF R M
+  | [], l => l.onScalar q(Neg.neg)
+  | l, [] => l
+  | ((a₁, x₁), k₁) :: t₁, ((a₂, x₂), k₂) :: t₂ =>
+    if k₁ < k₂ then
+      ((a₁, x₁), k₁) :: sub iR t₁ (((a₂, x₂), k₂) :: t₂)
+    else if k₁ = k₂ then
+      ((q($a₁ - $a₂), x₁), k₁) :: sub iR t₁ t₂
+    else
+      ((q(-$a₂), x₂), k₂) :: sub iR (((a₁, x₁), k₁) :: t₁) t₂
+
+/-- Given two terms `l₁`, `l₂` of type `qNF R M`, i.e. lists of `(Q($R) × Q($M)) × ℕ`s (two `Expr`s
+and a natural number), recursively construct a proof that in the `$R`-module `$M`, the difference
+of the "linear combinations" represented by `l₁` and `l₂` is the linear combination represented by
+`Module.qNF.sub iR l₁ l₁`.-/
+def mkSubProof (iR : Q(Ring $R)) (iM : Q(AddCommGroup $M)) (iRM : Q(Module $R $M))
+    (l₁ l₂ : qNF R M) :
+    Q(NF.eval $(l₁.toNF) - NF.eval $(l₂.toNF) = NF.eval $((qNF.sub iR l₁ l₂).toNF)) :=
+  match l₁, l₂ with
+  | [], l => (q(NF.zero_sub_eq_eval $(l.toNF)):)
+  | l, [] => (q(sub_zero (NF.eval $(l.toNF))):)
+  | ((a₁, x₁), k₁) :: t₁, ((a₂, x₂), k₂) :: t₂ =>
+    if k₁ < k₂ then
+      let pf := mkSubProof iR iM iRM t₁ (((a₂, x₂), k₂) :: t₂)
+      (q(NF.sub_eq_eval₁ ($a₁, $x₁) $pf):)
+    else if k₁ = k₂ then
+      let pf := mkSubProof iR iM iRM t₁ t₂
+      (q(NF.sub_eq_eval₂ $a₁ $a₂ $x₁ $pf):)
+    else
+      let pf := mkSubProof iR iM iRM (((a₁, x₁), k₁) :: t₁) t₂
+      (q(NF.sub_eq_eval₃ ($a₂, $x₂) $pf):)
+
+variable {iM : Q(AddCommMonoid $M)}
+  {u₁ : Level} {R₁ : Q(Type u₁)} {iR₁ : Q(Semiring $R₁)} (iRM₁ : Q(@Module $R₁ $M $iR₁ $iM))
+  {u₂ : Level} {R₂ : Q(Type u₂)} (iR₂ : Q(Semiring $R₂)) (iRM₂ : Q(@Module $R₂ $M $iR₂ $iM))
+
+/-- Given an expression `M` representing a type which is an `AddCommMonoid` and a module over *two*
+semirings `R₁` and `R₂`, find the "bigger" of the two semirings.  That is, we assume that it will
+turn out to be the case that either (1) `R₁` is an `R₂`-algebra and the `R₂` scalar action on `M` is
+induced from `R₁`'s scalar action on `M`, or (2) vice versa; we return the semiring `R₁` in the
+first case and `R₂` in the second case.
```

Comment:
Can you add a TODO comment here about supporting pairings like:
* `Int` and `NNRat` -> `Rat
* `Rat` and `NNReal` -> `Real`

Reply:
@eric-wieser I wanted to mention that, although on Zulip you suggested "hard coding those two latter cases and their permutations is probably enough in the medium term", I think the temptation to hard code these should be avoided, even as a stop-gap.  It's not hard to do it in proper generality -- [here](https://github.com/leanprover-community/mathlib4/commit/fe6b39553f566347e0c70b63a033c036c53e2058) is an implementation, which I am not adding to this PR because it would need some deduplication of code before being mathlib-ready.  But it can serve as a basis for a PR, if someone later needs this extension of scope.

Reply:
As you can see from that implementation, I don't think the TODO really belongs on this function (`matchRings`).  Rather, it belongs on the `parse` function, which already has a TODO comment along these lines.

Reply:
That PR looks like a great follow up, but the feature I was requesting here was automatically choosing the base ring in the examples above, in the cases where the user doesn't specify it.

Reply:
I was imagining registering the "union" types globally with a typeclass or an environment extension.

I'd settle for "hard-code the union type to use in the error message" in your follow-up PR (or a follow-up to it), if you feel strongly that any kind of heuristic for picking the union type is worse than forcing the user to choose.

Reply:
I see.  Your method would work, but it's a lot of overhead (and increase in difficulty of maintenance/improvement), and my guess is that this use case is very rare.  I'd prefer to do "hard-code the union type" for a while -- say, at least a year -- until we get a sense of how often this case turns up.

I should say frankly that (unless you consider it a condition of merging this PR) turning fe6b39553f566347e0c70b63a033c036c53e2058 into a follow-up PR will also be fairly low on my priority list, so it will be a while before we can make that experiment.

Reply:
> (and the error message only mentions the types the tactic already encountered).

I'm suggesting that the error message could suggest to the user what union type they should provide.

> I should say frankly that (unless you consider it a condition of merging this PR) turning https://github.com/leanprover-community/mathlib4/commit/fe6b39553f566347e0c70b63a033c036c53e2058 into a follow-up PR will also be fairly low on my priority list,

That's fine, thanks for making the commit!

Since you've done the work, I would suggest you create a draft PR from the change, and reference it in a TODO. That way, if anyone needs the feature, they find your version without accidentally implementing it from scratch!

Reply:
> Since you've done the work, I would suggest you create a draft PR from the change, and reference it in a TODO. That way, if anyone needs the feature, they find your version without accidentally implementing it from scratch!

#16984!

Reply:
> I'm suggesting that the error message could suggest to the user what union type they should provide.

I think I still don't understand here, sorry -- you're suggesting doing this:

> I was imagining registering the "union" types globally with a typeclass `SmallestCommonAlgebra R1 R2 (outParam A) extends SMulCommClass R1 R2 A` or an environment extension.

but using the "union-type lookup" *only* for error messages?